app/src/main/java/unipd/se18/ocrcamera/ResultActivity.java

```diff
@@ -0,0 +1,361 @@
+                        public void run() {
+                            mOCRTextView.setText(finalTextRecognized);
+                            tViewInci.setText(Html.fromHtml("Ingredienti trovati: "+inciDetectorEtichetta(finalTextRecognized)+""));
+                        }
+                    });
+                }
+                else
+                {
+                    final String finalTextRecognized = textRecognized;
+                    runOnUiThread(new Runnable() {
+                        @Override
+                        public void run() {
+                            mOCRTextView.setText(finalTextRecognized);
+                            tViewInci.setText(Html.fromHtml("Ingredienti trovati: "+inciDetectorEtichetta(finalTextRecognized)+""));
+                        }
+                    });
+                }
+            } else {
+                Log.e("NOT_FOUND", "photo not found");
+            }
+            return textRecognized;
+        }
+
+        @Override
+        protected void onPostExecute(String s) {
+            progressDialog.dismiss();
+            // Saving in the preferences
+            SharedPreferences sharedPref = getApplicationContext().getSharedPreferences("prefs", Context.MODE_PRIVATE);
+            SharedPreferences.Editor editor = sharedPref.edit();
+            editor.putString("text", s);
+            editor.apply();
+        }
+
+        @Override
+        protected void onPreExecute() {
+            progressDialog = ProgressDialog.show(ResultActivity.this,
+                    progressMessage,
+                    "");
+        }
+    }
+
+
+    public Bitmap rotateBitmap(Bitmap original, float degrees) {
+        int width = original.getWidth();
+        int height = original.getHeight();
+
+        Matrix matrix = new Matrix();
+        matrix.preRotate(degrees);
+
+        Bitmap rotatedBitmap = Bitmap.createBitmap(original, 0, 0, width, height, matrix, true);
+        Bitmap mutableBitmap = rotatedBitmap.copy(Bitmap.Config.ARGB_8888, true);
+        Canvas canvas = new Canvas(mutableBitmap);
+        canvas.drawBitmap(original, 5.0f, 0.0f, null);
+
+        return rotatedBitmap;
+    }
+
+    /**
+     * @author Giovanni Fasan(g1), Giovanni Piva(g1)
+     * @param text String in which you have to find the ingredients
+     * @return String with the list of ingredients found in the text
+     * Put in database csv: https://www.youtube.com/watch?v=i-TqNzUryn8
+     */
+    public String inciDetector(String text){
+        String inci = "";
+        try {
+            InputStream is = getResources().openRawResource(R.raw.database);    //InputStream from database.csv
+            BufferedReader reader = new BufferedReader(new InputStreamReader(is, Charset.forName("UTF-8")));    //database reader
+            String ingredient;  //String of inci ingredient
+            while ((ingredient=reader.readLine())!= null){      //defines ingredient as each row of the database
```

Comment:
In this way you scan all the database each time, this operation takes O(n) where n is the size of the database (plus the contains() cost). You can reduce the number of operations by splitting the text given in input into words and check for each word if there is an ingredient with this word in the database; if the ingredients are indexed this would take less as the search wouldn't be sequential.

Reply:
I know very well that this function has the very bad performance, I used it mainly as a base to write inciDetectorEtichetta